tests/suites/test_suite_psa_crypto.function

```diff
@@ -264,6 +264,445 @@ typedef enum {
+            {
+                part_length = data_part_len;
+            }
+
+            PSA_ASSERT( psa_aead_update( &operation,
+                                         ( input_data->x + part_offset ),
+                                         part_length, part_data,
+                                         part_data_size,
+                                         &output_part_length ) );
+
+            if( output_data && output_part_length )
+            {
+                memcpy( ( output_data + part_offset ), part_data,
+                        output_part_length );
+            }
+
+            part_offset += part_length;
+            output_length += output_part_length;
+        }
+    }
+    else
+    {
+        /* Pass whole data in one go */
+        PSA_ASSERT( psa_aead_update( &operation, input_data->x,
+                                     input_data->len, output_data,
+                                     output_size, &output_length ) );
+    }
+
+    PSA_ASSERT( psa_aead_finish( &operation, final_data,
+                                 finish_output_size,
+                                 &output_part_length,
+                                 tag_buffer, tag_length,
+                                 &tag_size ) );
+
+    if( output_data && output_part_length )
+    {
+        memcpy( ( output_data + output_length ), final_data,
+                output_part_length );
+    }
+
+    TEST_EQUAL( tag_length, tag_size );
+
+    output_length += output_part_length;
+
+    if( output_data && tag_length )
+    {
+        memcpy( ( output_data + output_length ), tag_buffer, tag_length );
+    }
+
+    output_length += tag_length;
+
+    /* For all currently defined algorithms, PSA_AEAD_ENCRYPT_OUTPUT_SIZE
+     * should be exact. */
+    TEST_EQUAL( output_length,
+                PSA_AEAD_ENCRYPT_OUTPUT_SIZE( key_type, alg,
+                                              input_data->len ) );
+    TEST_ASSERT( output_length <=
+                     PSA_AEAD_ENCRYPT_OUTPUT_MAX_SIZE( input_data->len ) );
+
+    ASSERT_COMPARE( expected_result->x, expected_result->len,
+                    output_data, output_length );
+
+exit:
+    psa_destroy_key( key );
+    psa_aead_abort( &operation );
+    mbedtls_free( output_data );
+    mbedtls_free( part_data );
+    mbedtls_free( final_data );
+    PSA_DONE( );
+
+    return( status );
+}
+
+void aead_multipart_decrypt_internal( int key_type_arg, data_t *key_data,
+                                      int alg_arg,
+                                      data_t *nonce,
+                                      data_t *additional_data,
+                                      int ad_part_len,
+                                      data_t *input_data,
+                                      int data_part_len,
+                                      int test_set_lengths_arg,
+                                      data_t *expected_data,
+                                      int expected_result_arg )
```

Comment:
The way `expected_result_arg` is used doesn't make sense. Looking at the test data, they come in three types: positive test cases where all the API functions should return `PSA_SUCCESS`, invalid-signature test cases where all the API functions should return `PSA_SUCCESS` except that `psa_aead_verify` should return `PSA_ERROR_INVALID_SIGNATURE`, and bad-case test cases where one of the preparation functions (`setup` or `set_nonce`, possibly in the future `set_lengths`) should fail. The code below doesn't quite test all this correctly. For the early failure test cases, it doesn't make sense to enumerate data lengths that won't even be reached; there should be completely separate test code for those tests. For the good and invalid-signature test cases, all that's needed is a boolean indicating whether to expect a valid signature.